### PR TITLE
Fix VectorTools::integrate_difference()

### PR DESCRIPTION
**Pull request by bangerth**
All sorts of bizarre problems I'm running into today:

The VectorTools::integrate_difference() function allows users
to provide a weight function that can also serve as a component mask
to select individual components of the solution vector for error
computation. For components not selected, such a mask would then
simply be zero.

In some cases, the solution vector contains NaN numbers, for example
when one uses the FE_FaceQ element for certain components of the
solution vector and uses a quadrature formula for error evaluation
that has quadrature points in the interior of the cell. For any
"regular" solution component for which the component mask has a zero
weight, the value of that component will be multiplied by zero and
consequently does not add anything to the error computation. However,
if the NaNs of a FE_FaceQ are multiplied with zero weights, the result
is still a NaN, and adding it to the values times weights of the other
components results in NaNs -- in effect rendering it impossible to get
any information out of the VectorTools::integrate_difference()
function if one of the finite elements involved is FE_FaceQ.

This is now fixed by simply skipping vector components for which the
weight vector is zero. This has the same result as before for all
"normal" situations, but also properly skips the NaN case outlined
above.

For reference: The way these NaNs get into the FEValues output is here:
  https://github.com/dealii/dealii/blob/master/include/deal.II/fe/fe_poly_face.templates.h#L86
I've updated this comment as part of the patch to explain some of the
consequences.

The problem first showed up in a program by @sophy1029 which pointed
me in the right direction with the fix.

### CODE REVIEW

File: tests/vector_tools/integrate_difference_02_nan.cc [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+  // happens to have signaling_nans in it
+  Vector<double> solution (dofh.n_dofs ());
+
+  Vector<double> cellwise_errors (tria.n_active_cells());
+  QIterated<dim> quadrature (QTrapez<1>(), 5);
+
+  const ComponentSelectFunction<dim> component_mask (1, 3);
+  VectorTools::integrate_difference (dofh,
+                                     solution,
+                                     Functions::ZeroFunction<dim>(dim),
```
Comment (jppelteret) [inline tests/vector_tools/integrate_difference_02_nan.cc]: Shouldn't this be your function `Ref`? Otherwise, I don't see where `Ref` is used?

**Comment by jppelteret** [comment]
Apart from my query about the test, the rest looks good to me!

File: tests/vector_tools/integrate_difference_02_nan.cc (hunk 1 above)
Comment (bangerth) [reply tests/vector_tools/integrate_difference_02_nan.cc]: Oh dear, what have I done. The test clearly doesn't make any sense any more...

I'll get back to you when I have that sorted out.

**Comment by bangerth** [comment]
I fixed the test now. This should be ready for review again.

**Comment by masterleinad** [comment]
I don't see a reason not to merge.

**Comment by bangerth** [comment]
@sophy1029 -- this is now finally merged into the main development sources. Maybe not as part of this semester's project (we have a workaround) but after the semester is over, we'll re-install deal.II for you with this fix in place and then you don't need the workaround any more.

**Pull request merged**